listenbrainz/db/stats.py

```diff
@@ -41,18 +42,23 @@ def insert_user_stats(user_id, artists, recordings, releases, artist_count):
              artist_count (int): the total number of artists listened to by the user
     """
 
+    month = datetime.now().month
```

Comment:
You have this code in three places, it should be moved into a utility so that it's only called once.
A warning: Dates are hard. really hard. I'd like to see some changes made here to be a little bit more careful about our calculations:
Call `datetime.now()` only once and use the same value to get both the year and the month
`datetime.now()` returns the _local_ time of the server. This could be any value depending on what the timezone of the server is set to. We should  always use UTC here - `datetime.utcnow()`
For a method of getting previous/next month, I'd prefer to use libraries that help manage these kinds of calculations. for example:

    >>> t = datetime.datetime.utcnow().replace(day=1)
    >>> t + dateutil.relativedelta.relativedelta(months=-1)
    datetime.datetime(2019, 1, 1, 9, 21, 27, 478427)
    >>> t + dateutil.relativedelta.relativedelta(months=-2)
   datetime.datetime(2018, 12, 1, 9, 21, 27, 478427)

Note that going back 2 months from feb will take us to 2018 automatically. The `relativedelta` method has some smart functionality, for example going back 1 month from March 31 will go to Feb 28 or 29 correctly, but in order to make the expected behaviour more explicit I use `.replace(day=1)` to set the date of the month to the 1st.

Dates are typically stored with a - as a separator, not a . like in your previous example. You should also 0-pad 1 digit months: 2019-02 (https://en.wikipedia.org/wiki/ISO_8601#Calendar_dates)

I'm not exactly sure why you want this date information, but it would probably be a good idea to pass this information into the `insert_user_stats` method as parameters. This would make it easier to test, and clearer to know where the values were calculated from. Do some form of these dates come from bigquery?

Reply:
No, these dates are not coming from bigquery. We are storing date information to know the data/stats belong to which month. We might want to implement a functionality in future that will allow users to retrieve data of any month/week and not just the previous month, in that case these dates would be useful.

Reply:
thanks for the clarification, I guess we should be talking about spark and not bigquery here, but I guess the result is the same.
It's unclear to me how this script is going to be run, so maybe I'm missing some understanding on exactly what time-frame this data covers, and what happens when you run the consumer:
 * How often is this script run?
 * At the point that this script is run, what data is in spark? Is it summary data for all users for a given month? Do we know what data has been calculated? For what users, and for what timeframe this data covers? I understand that this is summary data for an entire month, therefore I find it surprising that the month that the data covers isn't part of the information.
 * What happens if we run this script twice in a row? Will this override data for the last month or append the data to the table? Does this mean that it might be possible to have two rows containing summary data for a user for the same month?
 * What happens if we run it at 23:59 on the last day of the month and we continue running over the border of the month? This means that the "last month" for two subsequently inserted rows would be different, even though I suspect that they are actually for the same month.

I think that its probably necessary to have:
 * A sentinel table that keeps track of what month/user combinations have been inserted (this could be part of the `statistics.user` table, either by adding a date field or by creating an index into the date field of the inner json file.
* and/or: report the month that these statistics are for directly from spark so that there is no confusion as to what month they cover. Use this data to insert the stats into the database.

Reply:
- At present we decided to run the script monthly. Every month, previous month data of users would be exported from lb-server to playground where we will calculate stats of users and push them into rabbitmq. Then at lb-server end data will be consumed and fed into the database. Therefore, the script will run as many times as the playground producer runs which is monthly as of now.
- When this script runs, we have all the listen data of users for the previous month in spark. Though we will have data of months before the previous months (spark will basically store all the listen data) but we will be processing only the data for last month. Yes, the script at spark end tells us what data has been calculated. Currently, it is top 100 of artists, recordings, and releases.  I have modified the spark script, it will contain a key timestamp (which we changed to YearMonth) that will tell us the year and month in which the data was calculated. For eg, 'YearMonth' : '2019-01'. Also, the data in stored in json files like 1.json, 2.json which tells us the month and the main directory would be like 2019 which tells us the year. 
- If the script runs in a row, the data will be overridden. At any point we will only store data of the previous month of a user in the DB as of now. We won't have two rows for the same month. We can talk over if we need to keep record of every month's data in DB. But, I think this would un-necessarily flood the DB. If at any point the user wants to query data of some sepcific period we can queue the request and send it to spark. 
- There won't be two subsequent rows for a user as of now but this point is something which I was thinking about. Presently, if we calculate data at 23:59 and this is March and we keep calculating till the morning of 1st april the YearMonth would be "2019-03", a month before April but actually, it should be february. Presently, We assumed that the calculation would be done by the mid of a month but this is something that should be discussed upon. Thanks for pointing out :)

Reply:
The spark script reports the month for which these stats have been calculated. This info will be useful if we decide not to override the data, and store json data in db of months other than the previous month.

Reply:
> The spark script reports the month for which these stats have been calculated.

I think this is a key point. If you have the data for the month that you can read from spark, then use that! This will solve lots of issues about getting the correct value for the previous month when adding the data to the database.

Reply:
Yes sure :)